editor/scene_tree_dock.cpp

```diff
@@ -1684,20 +1684,39 @@ void SceneTreeDock::_create() {
 		}
 
 	} else if (current_option == TOOL_REPLACE) {
+
+		ScriptEditorDebugger *sed = ScriptEditor::get_singleton()->get_debugger();
 		List<Node *> selection = editor_selection->get_selected_node_list();
 		ERR_FAIL_COND(selection.size() <= 0);
+
+		editor_data->get_undo_redo().create_action(TTR("Change Node Type"));
 		for (List<Node *>::Element *E = selection.front(); E; E = E->next()) {
 			Node *n = E->get();
 			ERR_FAIL_COND(!n);
 
 			Object *c = create_dialog->instance_selected();
 
 			ERR_FAIL_COND(!c);
-			Node *newnode = Object::cast_to<Node>(c);
-			ERR_FAIL_COND(!newnode);
+			Node *new_node = Object::cast_to<Node>(c);
+			ERR_FAIL_COND(!new_node);
 
-			replace_node(n, newnode);
+			String path_to_n = edited_scene->get_path_to(n);
+			int n_pos_in_parent = n->get_position_in_parent();
+			String n_name = n->get_name();
+
+			editor_data->get_undo_redo().add_do_method(this, "replace_node", n, new_node);
+			editor_data->get_undo_redo().add_do_reference(new_node);
+			editor_data->get_undo_redo().add_do_method(sed, "live_debug_replace_node", path_to_n, new_node->get_class(), n_name);
+			editor_data->get_undo_redo().add_do_method(this, "memdelete", n);
```

Comment:
Not sure but `UndoRedo` should release the old node through reference passed by `add_do_reference/add_undo_reference` automatically if the undo history is full, I don't see using `do/redo` method for explicitly freeing a node anywhere throughout the engine, so perhaps they are redundant, and I'm not sure how do you keep/restore the node's state once you delete it.